contracts/Minter.sol

```diff
@@ -0,0 +1,174 @@
+  uint256 constant public cap = 3000000000000000000000000000;
+  /// @notice Maximum number of pools
+  uint256 constant public poolsCap = 100;
+  /// @notice Maximum amount per block to each pool
+  uint256 public maxToPoolPerBlock;
+
+  /// @notice Struct to store information about each pool
+  struct Pool {
+    address receiver;
+    uint256 amountPerBlock;
+  }
+  /// @notice Array to store all pools
+  Pool[] public pools;
+
+  /// @notice Emitted when mint() is called
+  event Mint(address indexed receiver, uint256 amount);
+  /// @notice Emitted when pool is added
+  event PoolAdded(address indexed newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is updated
+  event PoolUpdated(uint256 index, address indexed newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is removed
+  event PoolRemoved(uint256 index, address indexed receiver, uint256 amount);
+  /// @notice Emitted when admin address is updated
+  event UpdateAdmin(address indexed admin, address indexed newAdmin);
+  /// @notice Emitted when maxToPoolPerBlock is changed
+  event UpdateMaxToPoolPerBlock(uint256 oldMaxToPoolPerBlock, uint256 newMaxToPoolPerBlock);
+
+  /// @notice Modifier to allow only admin to call certain functions
+  modifier onlyAdmin(){
+    require(msg.sender == admin, 'SPSMinter: Only admin');
+    _;
+  }
+
+  /**
+   * @notice Constructor of new minter contract
+   * @param newToken Address of the token to mint
+   * @param startBlock Initial lastMint block
+   * @param newAdmin Initial admin address
+   * @param newMaxToPoolPerBlock Maximum amount minted per block
+   */
+  constructor(address newToken, uint256 startBlock, address newAdmin, uint256 newMaxToPoolPerBlock){
+    require(startBlock >= block.number, "SPSMinter: Start block must be above current block");
+    require(newToken != address(0), 'SPSMinter: Token cannot be address 0');
+    require(newAdmin != address(0), 'SPSMinter: Admin cannot be address 0');
+
+    token = IMintable(newToken);
+    lastMintBlock = startBlock;
+    admin = newAdmin;
+    maxToPoolPerBlock = newMaxToPoolPerBlock;
+
+    require(token.decimals() == 18, "SPSMinter: Token doesn't have 18 decimals");
+
+    emit UpdateAdmin(address(0), newAdmin);
+  }
+
+  /**
+   * @notice Mint tokens to all pools, can be called by anyone
+   */
+  function mint() public {
+    require(totalMinted != cap, "SPSMinter: Cap reached");
+    require(block.number > lastMintBlock, "SPSMinter: Mint block not yet reached");
+
+
+    uint256 mintDifference;
+    unchecked {
+      mintDifference = block.number - lastMintBlock;
+    }
+
+    lastMintBlock = block.number;
+
+    for (uint256 i = 0; i < pools.length; i++){
+      uint256 amount = pools[i].amountPerBlock * mintDifference;
+
+      if(totalMinted + amount >= cap){
+        amount = cap - totalMinted;
+      }
+
+      token.mint(pools[i].receiver, amount);
+      totalMinted = totalMinted + amount;
+
+      emit Mint(pools[i].receiver, amount);
```

Comment:
In the case that the cap has already been reached, you're going to waste a lot of gas, and emit a bunch of events, with 0 actual tokens being minted. Probably not a big deal, but just want to call it out.